mempool/cat/pool.go

```diff
@@ -0,0 +1,765 @@
+
+	txmp.updateMtx.Lock()
+	txmp.height = blockHeight
+	txmp.notifiedTxsAvailable = false
+
+	if newPreFn != nil {
+		txmp.preCheckFn = newPreFn
+	}
+	if newPostFn != nil {
+		txmp.postCheckFn = newPostFn
+	}
+	txmp.updateMtx.Unlock()
+
+	txmp.metrics.SuccessfulTxs.Add(float64(len(blockTxs)))
+	for _, tx := range blockTxs {
+		// Regardless of success, remove the transaction from the mempool.
+		txmp.removeTxByKey(tx.Key())
+	}
+
+	txmp.purgeExpiredTxs(blockHeight)
+
+	// If there any uncommitted transactions left in the mempool, we either
+	// initiate re-CheckTx per remaining transaction or notify that remaining
+	// transactions are left.
+	size := txmp.Size()
+	txmp.metrics.Size.Set(float64(size))
+	if size > 0 {
+		if txmp.config.Recheck {
+			txmp.recheckTransactions()
+		} else {
+			txmp.notifyTxsAvailable()
+		}
+	}
+	return nil
+}
+
+// addNewTransaction handles the ABCI CheckTx response for the first time a
+// transaction is added to the mempool.  A recheck after a block is committed
+// goes to handleRecheckResult.
+//
+// If either the application rejected the transaction or a post-check hook is
+// defined and rejects the transaction, it is discarded.
+//
+// Otherwise, if the mempool is full, check for lower-priority transactions
+// that can be evicted to make room for the new one. If no such transactions
+// exist, this transaction is logged and dropped; otherwise the selected
+// transactions are evicted.
+//
+// Finally, the new transaction is added and size stats updated.
+func (txmp *TxPool) addNewTransaction(wtx *wrappedTx, checkTxRes *abci.ResponseCheckTx) error {
+	// At this point the application has ruled the transaction valid, but the
+	// mempool might be full. If so, find the lowest-priority items with lower
+	// priority than the application assigned to this new one, and evict as many
+	// of them as necessary to make room for tx. If no such items exist, we
+	// discard tx.
+	if !txmp.canAddTx(wtx.size()) {
+		victims, victimBytes := txmp.store.getTxsBelowPriority(wtx.priority)
+
+		// If there are no suitable eviction candidates, or the total size of
+		// those candidates is not enough to make room for the new transaction,
+		// drop the new one.
+		if len(victims) == 0 || victimBytes < wtx.size() {
+			txmp.metrics.EvictedTxs.Add(1)
+			txmp.evictedTxs.Push(wtx)
+			checkTxRes.MempoolError = fmt.Sprintf("rejected valid incoming transaction; mempool is full (%X)",
+				wtx.key)
+			return fmt.Errorf("rejected valid incoming transaction; mempool is full (%X). Size: (%d:%d)",
+				wtx.key.String(), txmp.Size(), txmp.SizeBytes())
+		}
+
+		txmp.logger.Debug("evicting lower-priority transactions",
+			"new_tx", wtx.key.String(),
+			"new_priority", wtx.priority,
+		)
+
+		// Sort lowest priority items first so they will be evicted first.  Break
+		// ties in favor of newer items (to maintain FIFO semantics in a group).
+		sort.Slice(victims, func(i, j int) bool {
```

Comment:
any chance we can refactor this sorting? since we are sorting by the same criteria in multiple places, would be super annoying if we accidentally changed one and not the other.

Reply:
Yeah we could factor this out. It's only called in two places and it's not required that they use the same criteria i.e. eviction order and staging order could be different.